### ISSUE TITLE
ImportError: DLL load failed while importing ibm_db:

### ISSUE THREAD

**Issue by ZephyrusMB** [ibmdb/python-ibmdb#887]
"ImportError: DLL load failed while importing ibm_db" raised while executing script. Working with ibm_db version 3.1.4.
Dlls are not present with 3.2.0 version.

- Operating System Name: Microsoft Windows 11 10.0.22621
- Python Version: 3.11.4 and 3.11.5
- ibm_db version: 3.2.0

- Test script to reproduce the problem
Failed at line 1 `import ibm_db`
```
import ibm_db
import ibm_db_dbi
ibm_db_conn= ibm_db.connect(connectionString, "", "")
conn = ibm_db_dbi.Connection(ibm_db_conn)
```

Steps to Reproduce:

1. install module via `pip install --upgrade ibm_db`
2. execute script python shared in the post

**Comment by Earammak** [comment]
@ZephyrusMB, Can you please run as  below, before import ibm_db.
import os
os.add_dll_directory('path to clidriver installation until bin')
import ibm_db

**Comment by dverweij-dev** [comment]
Also upgraded from ibm-db version 3.1.4 to 3.2.0 and ran into the error "ImportError: DLL load failed while importing ibm_db".
OS = Windows 10 Enterprise and version = 10.0.19044.
And yes, when adding os.add_dll_directory('path to clidriver installation until bin') the problem is fixed.
Before changing things, can I assume this is a bug/omission which wil be fixed in the next version of this package?

**Comment by Earammak** [comment]
@ZephyrusMB This we need to add manually for windows only. And the same we documented in README.

**Comment by dverweij-dev** [comment]
I studied the README before posting this comment, so it was clear to me what needed to be done. The issue I want to address is that this package is not downwards compatible and platform agnostic anymore and we need to change several scripts. 

Using ibm-db==3.1.4 we can run the same scripts on Windows, Linux and in Containers.

If I want to use ibm-db==3.2.0 I have to alter these scripts to check if os.name == 'nt' (or os.name != 'posix') and then execute os.add_dll_directory(os.getenv('IBM_DB_HOME') + '/bin').

I'm using this driver mainly in combination with RobotFramework (this is a testautomation framework) third party libraries (e.g. robotframework-databaselibrary). I don't want to change these third party libraries, so need to use IF and Evaluate construction to make it platform independent in my testscripts again:

```
Connect To DB2 Database
    # Fix DLL issue with package ibm-db version 3.2.0
    IF    (os.name == 'nt')
        Evaluate    os.add_dll_directory(os.getenv('IBM_DB_HOME') + '/bin')    modules=os
    END
    ...
    Connect To Database    ibm_db_dbi    ${DB2_DATABASE}    ${username}    ${password}    ${DB2_HOST}    ${DB2_PORT}
    ...
    ...
```
The question is, can't you add this construction:

```
if os.name == 'nt':
    os.add_dll_directory(os.getenv('IBM_DB_HOME') + '/bin')
```

or other construction to the ibm-db module to make it platform agnostic again and downwards compatible with ibm-db==3.1.4?

If there is good reason to not do this, please shed some light on this as to why. Is it conflicting somewhere else?

If it cannot be changed, then we will stick to ibm-db==3.1.4 for now an change our scripts later on to work with ibm-db=3.2.0 or later version.

**Comment by Rishi-spec-art** [comment]
@Earammak I'm getting this error: 
OSError: [WinError 87] The parameter is incorrect: 'path to clidriver installation until bin'

**Comment by ZephyrusMB** [comment]
Yes please as described by @dverweij-dev the expected behavior is to be platform agnostic as in version 3.1.4.
Is there any specific reason to introduce this breaking change into version 3.2.0? Would you solve this point in the future? 

Kind regards

**Comment by frbelotto** [comment]
Same issue here. 

```
`---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
[c:\Users\F3164582\Banco](file:///C:/Users/F3164582/Banco) do Brasil S.A\DINED Biz Dev - General\Python\LojaBB\Campanhas.ipynb Cell 73 line 4
      [2](vscode-notebook-cell:/c%3A/Users/F3164582/Banco%20do%20Brasil%20S.A/DINED%20Biz%20Dev%20-%20General/Python/LojaBB/Campanhas.ipynb#Y142sZmlsZQ%3D%3D?line=1) usr = 'DB2I023A'
      [3](vscode-notebook-cell:/c%3A/Users/F3164582/Banco%20do%20Brasil%20S.A/DINED%20Biz%20Dev%20-%20General/Python/LojaBB/Campanhas.ipynb#Y142sZmlsZQ%3D%3D?line=2) pwd = '31645828'
----> [4](vscode-notebook-cell:/c%3A/Users/F3164582/Banco%20do%20Brasil%20S.A/DINED%20Biz%20Dev%20-%20General/Python/LojaBB/Campanhas.ipynb#Y142sZmlsZQ%3D%3D?line=3) engine = Conectar(usr, pwd)
      [5](vscode-notebook-cell:/c%3A/Users/F3164582/Banco%20do%20Brasil%20S.A/DINED%20Biz%20Dev%20-%20General/Python/LojaBB/Campanhas.ipynb#Y142sZmlsZQ%3D%3D?line=4) conn = engine.engine.connect()

File [c:\Users\F3164582\Banco](file:///C:/Users/F3164582/Banco) do Brasil S.A\DINED Biz Dev - General\Python\LojaBB\DB2\ConexaoDB2.py:9, in Conectar(usr, pwd)
      8 def Conectar(usr, pwd):
----> 9     engine =  create_engine(f'ibm_db_sa://{usr}:{pwd}@gwdb2.bb.com.br:50100/BDB2P04;', echo=False)
     10     engine = engine.connect().execution_options(stream_results=True)
     11     return engine

File <string>:2, in create_engine(url, **kwargs)

File [c:\Users\F3164582\AppData\Local\Programs\Python\Python311\Lib\site-packages\sqlalchemy\util\deprecations.py:281](file:///C:/Users/F3164582/AppData/Local/Programs/Python/Python311/Lib/site-packages/sqlalchemy/util/deprecations.py:281), in deprecated_params.<locals>.decorate.<locals>.warned(fn, *args, **kwargs)
    274     if m in kwargs:
    275         _warn_with_version(
    276             messages[m],
    277             versions[m],
    278             version_warnings[m],
    279             stacklevel=3,
    280         )
--> 281 return fn(*args, **kwargs)
...
---> 45 import ibm_db
     46 __version__ = ibm_db.__version__
     48 # Constants for specifying database connection options.

ImportError: DLL load failed while importing ibm_db: Não foi possível encontrar o módulo especificado.`
```

**Comment by bimalkjha** [comment]
@frbelotto You need to run below command as documented [here](https://github.com/ibmdb/python-ibmdb#installation) as Note:
```
import os
os.add_dll_directory('path to clidriver installation until bin')
import ibm_db
```
Also, this solution is already given in this comment: https://github.com/ibmdb/python-ibmdb/issues/887#issuecomment-1709437336
Probably, you have not read the full thread. Try it. Thanks.

**Comment by readytheory** [comment]
> @Earammak I'm getting this error: OSError: [WinError 87] The parameter is incorrect: 'path to clidriver installation until bin'

@Rishi-spec-art they mean something like:

> os.add_dll_directory('/venv/cashapp/Lib/site-packages/clidriver/bin')

... whatever the path is to where pip put your clidriver/bin.   IBM team, I'd suggest clarifying that wording.  "until bin" sounds to me like you don't want it to stop at "clidrivder/"

**Comment by frbelotto** [comment]
> @frbelotto You need to run below command as documented [here](https://github.com/ibmdb/python-ibmdb#installation) as Note:
> 
> ```
> import os
> os.add_dll_directory('path to clidriver installation until bin')
> import ibm_db
> ```
> 
> Also, this solution is already given in this comment: [#887 (comment)](https://github.com/ibmdb/python-ibmdb/issues/887#issuecomment-1709437336) Probably, you have not read the full thread. Try it. Thanks.

Just to understand. Such installation (ODBC and CLI driver) wasn't needed before? I have been using IBM_DB for a while and started getting this error recently.

**Comment by bimalkjha** [comment]
@readytheory Yes, do not stop at clidriver but clidriver/bin as in `os.add_dll_directory('/venv/cashapp/Lib/site-packages/clidriver/bin')`. Thanks.

**Comment by bimalkjha** [comment]
@frbelotto You are right. Actually from ibm_db version 3.2.0, we have discontinued shipping ibm_db.dll file and switched on wheel support for Linux, Mac and Windows to avoid compilation related issues. So, change in behavior. Thanks.

**Comment by LinusConradM** [comment]
I got this error despite following the guide
1. install module via pip install --upgrade ibm_db (this shows **"Requirement already satisfied: ibm_db in c:\users\cmuhirwe\appdata\local\anaconda3\lib\site-packages (3.2.0)"**

2. execute script python shared in the post 
import ibm_db
import ibm_db_dbi
ibm_db_conn= ibm_db.connect(connectionString, "", "")
conn = ibm_db_dbi.Connection(ibm_db_conn)



Bellow is the error 
ImportError                               Traceback (most recent call last)
Cell In[18], line 1
----> 1 import ibm_db
      2 import ibm_db_dbi
      3 ibm_db_conn= ibm_db.connect(connectionString, "", "")

ImportError: DLL load failed while importing ibm_db: The specified module could not be found.

**Comment by bimalkjha** [comment]
@LinusConradM Seems you have not went through the comments in this issue. Please check it: https://github.com/ibmdb/python-ibmdb/issues/887#issuecomment-1766239125 Thanks.

**Comment by rredburn** [comment]
add_dll_directory to every script that needs db2 access?  What a terrible user experience, I can't believe this is the best answer.  Please re-evaluate the strategy.

**Comment by ditarnnt** [comment]
this is really helpfull, thanks

**Comment by frbelotto** [comment]
Does anyone can confirm what is the lastest version of the library that does not request adding the dll manually? I want to set it up on my requests.txt.

**Comment by bchoudhary6415** [comment]
@frbelotto python-ibm_db v3.1.4 is the latest version of the library that does not request adding the dll manually.

Also from python version
The os.add_dll_directory() function is part of the os module in Python (introduced in Python 3.8), and it is used to add a directory to the DLL (Dynamic Link Library) search path. This function is particularly useful when working with third-party DLLs or shared libraries that your Python application depends on.

Before Python 3.8, the DLL search path was influenced by the PATH environment variable. However, this could lead to issues, especially when Python was embedded in larger applications. The os.add_dll_directory() function was introduced to provide a more controlled and isolated way of managing the DLL search path.

Here's a basic example of how to use os.add_dll_directory():

import os

Add a directory to the DLL search path
dll_directory = r'C:\Path\To\Your\DLLs'
os.add_dll_directory(dll_directory)

Now, when loading a DLL or using ctypes, the specified directory is considered in the search path
This function is particularly useful when dealing with scenarios where the DLLs needed by your Python application are not in the system's default search paths. By using os.add_dll_directory(), you can ensure that the specified directory is considered when resolving DLL dependencies.

It's important to note that this function is available only in Python 3.8 and later versions. If you're working with an earlier version of Python, you might need to consider alternative approaches or upgrade to a newer Python version.

Thanks.

**Comment by Xnot** [comment]
Regardless of whether you end up adding the ddl back or not, could you consider issuing deprecation warnings for breaking changes in the future?

**Comment by ralberrto** [comment]
I'm running into this issue on redhat:

```
>>> ibm_db.createdb(conn, "NEWDB")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
Exception: Error in loading db2cli.dll library file
```

Am I missing something here?

**Comment by kxmatjob** [comment]
***********************************************************************************
I have found a way to place the clidriver and ibm_db_dlls folders in the ibm_db-3.1.4-py3.8.egg directory, as well as ibm_db-py and ibm_db_dbi.py, in the current program directory. 
***********************************************************************************

**Comment by ksaket1998** [comment]
> @ZephyrusMB, Can you please run as below, before import ibm_db. import os os.add_dll_directory('path to clidriver installation until bin') import ibm_db

This works fine, can connect to db2 after running os.add_dll_directory. 
But can you help me how to make it work after converting the python script to an exe which would be used in multiple systems ?

**Comment by bimalkjha** [comment]
@ksaket1998 In your python script, you need to code to find the correct `site_packages` path that contains ibm_db.* file using code similar to below one:
```
import sys
for path in sys.path:
    if 'site-packages' in path:
        print(path)
```
Then check for clidriver under site_packages or directly check for clidriver instead of ibm_db* file. Then construct full path till clidriver\bin and add it to os.add_dll_directory(). Thanks.

**Comment by EliezerRamirezRuiz** [comment]
hey buddy, how are you?
I am recently developing a script, using the `ibm_db` library and I got the same error as you.

I have found a [solution that at least it has helped me and other users](https://stackoverflow.com/questions/77139499/ibm-db-3-2-0-importerror-dll-load-failed-while-importing-ibm-db-the-specified), was to uninstall the library after the driver and do the step by step again. 

1.- Install the IBM Driver
2.- Place a variable in the system called `IBM_DB_HOME`.
3.- Reinstall the `ibm_db` library.
4.- If it doesn't work try the following

```PYTHON
from os import add_dll_directory
from os import environ


if "IBM_DB_HOME" not in environ["PATH"]:
    add_dll_directory(r"path ibm drivers")

from ibm_db import connect
from ..enviroments import P8CMPROD

p8cmprod_connection = connect("dsn", "user", "password")
```

I have also seen the solution of mapping to the clidriver that comes with the library but at least it didn't work for me, so any solution that works for you would be good to know.

**Comment by bimalkjha** [comment]
@frbelotto To install or update python ibm_db driver, no need to install DB2 Client. It automatically downloads and install clidriver internally. But, if you have requirement to use specific version of Db2 Client or driver and want to avoid automatic download of clidriver during installation of ibm_db, then you need to follow the requirement of installation of Db2 Client.. and yes, most of the website wants registration to know their users. 
Also, the wheel package of ibm_db comes with clidriver in-built. Thanks.

**Comment by frbelotto** [comment]
> @frbelotto To install or update python ibm_db driver, no need to install DB2 Client. It automatically downloads and install clidriver internally. But, if you have requirement to use specific version of Db2 Client or driver and want to avoid automatic download of clidriver during installation of ibm_db, then you need to follow the requirement of installation of Db2 Client.. and yes, most of the website wants registration to know their users. 
> 
> Also, the wheel package of ibm_db comes with clidriver in-built. Search about IBM_DB_HOME  on https://github.com/ibmdb/python-ibmdb page to know more about download of db2client by ibm_db and installation steps using separate db2client install. Thanks.

Hello! 
The ibm db installation,on windows , just talk about setting the environment path, but it says "if you have db2 cli installed " 

I've tried to install ibm db today, but as expected it was complaining about a missing DLL. 

I don't need to use any specific version. I am trying to use a newer version so I could start using Python 3.12

**Comment by bimalkjha** [comment]
@frbelotto This missing DLL error on windows is a known issue and we have already documented the solution and updated on [this comment](https://github.com/ibmdb/python-ibmdb/issues/887#issuecomment-1766239125) too. Thing is the path of clidriver/bin directory that you need to pass to `add_dll_directory()` is from your site_packages path. If you check your `site_packages` directory of python, you can see clidriver is present there. Just add the full path of clidriver\bin directory to `add_dll_directory()`. No need to have separated installation of db2client to get this directory. Thanks.

**Comment by bimalkjha** [comment]
Below screenshot show how it works without installing separate db2client in the system post pip install:

![image](https://github.com/user-attachments/assets/f07910ab-5290-4507-9a0c-83d5306b65b2)

**Comment by Gamhigh** [comment]
![Screenshot 2024-10-22 081618](https://github.com/user-attachments/assets/97d343fa-0517-48b0-8d50-97c8b8b6b21c)
i'm setting IBM DB2 in D\ , and i can't connect with the ibm because DLL
    import ibm_db
ImportError: DLL load failed while importing ibm_db: The specified module could not be found.

**Comment by Gamhigh** [comment]
@bchoudhary6415 thank you very much, I was able to log in. I know it's fake password

**Comment by klaasbrant** [comment]
What a mess... At the moment you cannot test properly with venv or uv.  This problem is not unique, there are other packages that have the same problem and solved it! 

Please have a look at psycopg2. How it nicely escapes the dll hell. Please, please fix this, And no we don't want to compile the client, just distribute the dll like psycopg2 does!

**Comment by klaasbrant** [comment]
OK instead of you complaining, let me offer a solution. I see the first step is already made. The cli drivers are installed with ibm_db.
So all we need is a bit of code that can be part of the ibm_db package. Like this:

import os
import site

# The problem is unique to windows
if os.name == "nt":
    found = False
    # Did the customer install the clidriver? If so the IBM_DB_HOME environment variable should be set
    if os.environ.get("IBM_DB_HOME"):
        clidriver_path = os.environ["IBM_DB_HOME"]
        if os.path.exists(clidriver_path):
            os.add_dll_directory(os.path.join(clidriver_path, "bin"))
            found = True
            print(f"Added {clidriver_path} to DLL search path")

    # If not, we will search the site-packages for the clidriver that came with ibm_db
    else:
        for site_packages in site.getsitepackages(): #careful there are multiple locations!
            possible_clidriver_folder = os.path.join(site_packages, "clidriver")
            if os.path.exists(possible_clidriver_folder):
                os.add_dll_directory(os.path.join(possible_clidriver_folder, "bin"))
                found
                print(f"Added {possible_clidriver_folder} to DLL search path")
                found = True
                break

    if not found:
        raise FileNotFoundError("clidriver folder not found, you cannot use ibm_db")

The code executes only on windows.  First it looks if the environment var IBM_DB_HOME exists. If so the customer has installed the clidriver and we use that one. If not it should be able to find the cli that came with the install of ibm_db and use that one. If all fails we raise a filenotfound error. 

That was easy!

p.s. we don't need the print statements in the final version :-)
